torch/distributed/optim/apply_optimizer_in_backward.py

```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Any, Dict, Iterable, Type
+
+import torch
+
+__all__ = []  # type: ignore[var-annotated]
+
+
+def apply_optimizer_in_backward(
+    optimizer_class: Type[torch.optim.Optimizer],
+    params: Iterable[torch.nn.Parameter],
+    optimizer_kwargs: Dict[str, Any],
+) -> None:
+    """
+    Upon backwards(), parameters will fire the corresponding optimizer
+    Each parameter will have the optimizer_class and optimizer_kwargs attached to
+    _optimizer and _optimizer_kwargs.
+
+    Note - gradients for these parameters will be set to None after backwards().
+    This means that any other (non applied) optimizer over this parameter will be
+    a no-op.
+
+    Args:
+        optimizer_class: Type[torch.optim.Optimizer]: Optimizer to apply to parameter
+        params: Iterator[nn.Parameter]: parameters to apply optimizer state to
+        optimizer_kwargs: Dict[str, Any]: kwargs to pass to optimizer constructor
+
+    Example::
+        params_generator = model.parameters()
+        param_1 = next(params_generator)
+        param_2 = list(params_generator)
+
+        apply_optimizer_in_backward(torch.optim.SGD, [param_1], {"lr": .02})
+        apply_optimizer_in_backward(torch.optim.Adam, param_2, {"lr": .04})
+
+        print(param_1._optimizer, param_1._optimizer_kwargs)
+        >> torch.optim.SGD, {"lr": .02}
+    """
+
+    def _apply_optimizer_in_backward_to_param(param: torch.nn.Parameter) -> None:
+        # acc_grad creates a new node in the auto_grad graph that comes after
+        # the parameter node. In this node's backwards we call the overlapped
+        # optimizer.step(). We cannot do the backwards hook on param because
```

Comment:
I think what this is trying to explain is that we cannot register the backward hook on the parameter itself. It must be on the parameter's `AccumulateGrad` object.

This is indeed my understanding, but I think this can be stated more clearly. The point is that registering a hook on the parameter means that hook runs before its gradient computation, while registering a hook on the `AccumulateGrad` object means that hook runs after the gradient computation.

Reply:
I clarified this in next update.